File: python/array/leetcode/contains_duplicate_ii.py

# 是否存在重复元素II
from typing import List


class Solution:

    # 暴力法，依次遍历
    def containsNearbyDuplicate_1(self, nums: List[int], k: int) -> bool:
        for i in range(len(nums) - 1):
            for j in range(i + 1, len(nums)):
                if j - i > k:
                    break
                if nums[j] == nums[i]:
                    return True
        return False

    # 使用一个哈希表
    def containsNearbyDuplicate_2(self, nums: List[int], k: int) -> bool:
        m = {}
        for i, v in enumerate(nums):
            if v in m and i - m[v] <= k:
                return True
            m[v] = i
        return False

File: go/array/leetcode/contains_duplicate_ii.go
package main

//存在重复元素II

//遍历查找
func containsNearbyDuplicate1(nums []int, k int) bool {
	for i := 0; i < len(nums)-1; i++ {
		for j := i + 1; j < len(nums); j++ {
			if nums[i] == nums[j] && j-i <= k {
				return true
			}
		}
	}
	return false
}

//使用一个哈希表
func containsNearbyDuplicate2(nums []int, k int) bool {
	m := make(map[int]int)
	for i, n := range nums {
		if v, ok := m[n]; ok && i-v <= k {
			return true
		}
		m[n] = i
	}
	return false
}
